add FilterFeatures input transform and tests

jduerholt opened the pull request:
<!--
Thank you for sending the PR! We appreciate you spending the time to make BoTorch better.

Help us understand your motivation by explaining why you decided to make this change.

You can learn more about contributing to BoTorch here: https://github.com/pytorch/botorch/blob/main/CONTRIBUTING.md
-->

## Motivation

This PR adds the possibility for using GPs in a `ModelListGP` sharing only subsets of features. This is useful in multiobjective optimization if it is known that a feature is not important for one objective but has to be taken into account for another objective.

It is implemented by adding a new input transform called `FilterFeatures`.

Addresses https://github.com/pytorch/botorch/issues/1028

### Have you read the [Contributing Guidelines on pull requests](https://github.com/pytorch/botorch/blob/main/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

I added tests for the new transform according to the tests implemented for the other input transforms.

## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/pytorch/botorch, and link to your PR here.)

Balandat commented on botorch/models/transforms/input.py:
```diff
@@ -836,6 +836,78 @@ def transform(self, X: Tensor) -> Tensor:
+            raise ValueError("`feature_indices` tensor must be int64/long!")
+        if (feature_indices < 0).any():
+            raise ValueError(
+                "Elements of `feature_indices` have to larger/equal to zero!"
```
```suggestion
                "Elements of `feature_indices` have to be larger/equal to zero!"
```

Balandat commented on botorch/models/transforms/input.py:
```diff
@@ -836,6 +836,78 @@ def transform(self, X: Tensor) -> Tensor:
+        """
+        super().__init__()
+        if feature_indices.dim() != 1:
+            raise ValueError("`feature_indices` must be a onedimensional tensor!")
```
```suggestion
            raise ValueError("`feature_indices` must be a one-dimensional tensor!")
```

Balandat commented on botorch/models/transforms/input.py:
```diff
@@ -836,6 +836,78 @@ def transform(self, X: Tensor) -> Tensor:
+        Returns:
+            A boolean indicating if the other transform is equivalent.
+        """
+        return (
```
Could you add a test for this as well, please?

jduerholt replied:
I added the test and fixed a bug in the equals method.

Balandat commented on botorch/models/transforms/input.py:
```diff
@@ -836,6 +836,78 @@ def transform(self, X: Tensor) -> Tensor:
+            X: A `batch_shape x q x d`-dim tensor of inputs.
+
+        Returns:
+            A `batch_shape x q x e`-dim tensor of filtered inputs, where is the length of `feature_indices`.
```
this line will probably be too long for flake?

jduerholt replied:
Fixed also this.

Balandat reviewed (commented):
Thanks for putting this up, looks great. There is one little test to add, and then code formatting is failing, likely b/c this is based on some old base commit. The easiest way to fix this is probably to merge `main` back in there and make make sure to use the code formatter versions as they will be pinned in #1046.

jduerholt commented:
Hi @Balandat,

thanks for reviewing. I addressed your comments. I am not so fluent with all these formatters. I hope that it passes now ;)

Best,

Johannes

Balandat commented:
All good, thank you! Will get this landed soon.
